sql/driver/conn.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package driver
+
+import (
+	"bytes"
+	"database/sql/driver"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/structured"
+)
+
+// TODO(pmattis):
+//
+// - This file contains the experimental Cockroach sql driver. The driver
+//   currently parses SQL and executes key/value operations in order to execute
+//   the SQL. The execution will fairly quickly migrate to the server with the
+//   driver performing RPCs.
+//
+// - Flesh out basic insert, update, delete and select operations.
+//
+// - Figure out transaction story.
+
+type conn struct {
+	db       *client.DB
+	database string
+}
+
+func (c *conn) Close() error {
+	return nil
+}
+
+func (c *conn) Prepare(query string) (driver.Stmt, error) {
+	s, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	return &stmt{conn: c, stmt: s}, nil
+}
+
+func (c *conn) Exec(query string, args []driver.Value) (driver.Result, error) {
+	stmt, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	return c.exec(stmt, args)
+}
+
+func (c *conn) Query(query string, args []driver.Value) (driver.Rows, error) {
+	stmt, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	return c.query(stmt, args)
+}
+
+func (c *conn) Begin() (driver.Tx, error) {
+	return &tx{conn: c}, nil
+}
+
+func (c *conn) exec(stmt parser.Statement, args []driver.Value) (*result, error) {
+	rows, err := c.query(stmt, args)
+	if err != nil {
+		return nil, err
+	}
+	return &result{rowsAffected: int64(len(rows.rows))}, nil
+}
+
+func (c *conn) query(stmt parser.Statement, args []driver.Value) (*rows, error) {
+	switch p := stmt.(type) {
+	case *parser.CreateDatabase:
+		return c.createDatabase(p, args)
+	case *parser.CreateTable:
+		return c.createTable(p, args)
+	case *parser.Delete:
+		return c.del(p, args)
+	case *parser.Insert:
+		return c.insert(p, args)
+	case *parser.Select:
+		return c.sel(p, args)
```

Comment:
`{insert,update,delete}` can all operate on multiple rows. Should those be pluralized as well? I could also just capitalize the method names.